### ISSUE TITLE
Preserve line endings when repository is cloned

### ISSUE THREAD

**Issue by bburton** [LibreTime/libretime#141]
## Issue

When checking out the repository under Windows using the official Git version for Windows, it tries to be helpful by converting line endings to the Windows native CR+LF for all text files. Normally, this might not be an issue but when provisioning a VM with Vagrant, it breaks the install as Bash can't execute scripts with CR+LF line endings.

## Expected behavior

Git cloning under Windows preserves line endings of relevant scripts and files regardless of a user's local Git configuration so it's possible to provision a VM with Vagrant, install and run LibreTime in place without issue.

## Possible solution 1

Modify repository configuration:

`git config core.autocrlf input`

However, I'm not sure if this configuration change will persist so that it works when cloning remotely.

## Possible solution 2

Create a .gitattributes file in the root of the repository that is configured to preserve all line endings for any scripts or files required to provision a VM, install and execute LibreTime.  In this case, it appears each text file type that needs to preserve the LF line ending would have to be identified as "text eol=lf"

In this case, it can be helpful to identify every unique file extension in the repository. The following command line does this and also lists every file with no extension when run from the root folder of the Git workspace:

`git ls-tree -r HEAD --name-only | perl -ne 'print $1 if m/\.?([^.\/]+)$/' | sort -u`

Attached is a proposed .gitattributes file.
[gitattributes.txt](https://github.com/LibreTime/libretime/files/891680/gitattributes.txt)

## Possible solution 3

Don't make any changes other than to document the workaround below.

## References

[GitHub Help - Dealing with line endings](https://help.github.com/articles/dealing-with-line-endings/#platform-linux)

## Workaround

Before cloning the repository, change the global setting for `core.autocrlf` to `input`. This will preserve the line endings when files are checked out into the workspace after cloning. However, any new text files must be created with LF line endings as the CR+LF will not be converted to LF on commit:

`git config --global core.autocrlf input`

Then perform the clone.

**Comment by hairmare** [comment]
I haven't had to use Windows in ages so I can't say if this isn't something special that every windows git user need to be aware of.

Solution 1 is only local to the repo and would need changing for each new clone. I feel the workaround is a bit much for someone who wants to get into LibreTime and it also kinda pollutes the global git config. 

As long as the gitattributes change doesn't have any side-effects on Unix platforms I'd be happy to merge a PR containing your documented fix for windows. Does it really need all the files to be listed? Shouldn't a double star glob (`**`) catch them all?

**Comment by hairmare** [comment]
@bburton I just sent you an invite to the Contributors team so you can tag and assign issues.

**Comment by bburton** [comment]
@hairmare Thanks for that!

This isn't the most typical cross-platform project with just source code as even though it can be checked out under Windows, it needs at the same time to be executable under Linux so the normal rules for determining line ending conversions don't apply. 

Yes, the workaround is problematic and possibly error-prone for the user as it could mean switching the global config depending on what repo is being worked on. It would be very easy to forget to switch it appropriately resulting in problems.

As far as globs go, a single star is already recursive. The following examples apply to the entire tree:
```
* text=auto
*.sh text eol=lf
```

After some more investigation, I found it's possible to see how the .gitattributes file affects files and to see to which files a pattern applies by using the `git check-attr` command. The following command line will show how attributes are set for all files when run from the root folder in the workspace:
```
git ls-files |git check-attr --stdin -a |less
```

After modifying a .png file in the working directory and performing some tests, I was able to conclusively show that with the pattern `* text eol=lf`:
* Unconditionally marks ALL files as text - even binary files.
* Ensures newlines are normalized to LF.

The file facebook_widget1.png is modified in airtime_mvc/public/images/doc:

```
$ git ls-files |git check-attr --stdin -a
facebook_widget1.png: text: set
facebook_widget1.png: eol: lf
facebook_widget2.png: text: set
facebook_widget2.png: eol: lf

$ git diff facebook_widget1.png
diff --git a/airtime_mvc/public/images/doc/facebook_widget1.png b/airtime_mvc/public/images/doc/facebook_widget1.png
index 6b4a8cb..66097ad 100644
Binary files a/airtime_mvc/public/images/doc/facebook_widget1.png and b/airtime_mvc/public/images/doc/facebook_widget1.png differ
warning: CRLF will be replaced by LF in airtime_mvc/public/images/doc/facebook_widget1.png.
The file will have its original line endings in your working directory.
```

That's no good as .png's could be corrupted. Of course, adding the pattern `*.png binary` would fix this but it means all binary files must be explicitly specified as the automatic text vs. binary detection is overridden.

The pattern `* text=auto` uses Git's automatic binary file detection but it also translates newlines to CR+LF on Windows which is not what we want. After some more experimentation, I found it's possible to use the `eol=lf` modifier in conjunction with `text=auto` as in `text=auto eol=lf`. This is not documented at least in the [official gitattributes page](https://git-scm.com/docs/gitattributes).

So, given the following .gitattributes file:
```
* text=auto eol=lf
```

```
$ git ls-files |git check-attr --stdin -a
facebook_widget1.png: text: auto
facebook_widget1.png: eol: lf
facebook_widget2.png: text: auto
facebook_widget2.png: eol: lf

$ git diff facebook_widget1.png
diff --git a/airtime_mvc/public/images/doc/facebook_widget1.png b/airtime_mvc/public/images/doc/facebook_widget1.png
index 6b4a8cb..a8e4b94 100644
Binary files a/airtime_mvc/public/images/doc/facebook_widget1.png and b/airtime_mvc/public/images/doc/facebook_widget1.png differ
```

That's much better as no newline conversion is implied.

Adding the attribute `*.png binary` doesn't have any effect on the results of the diff output.

In addition, if an attempt is made to add a file with CR+LF newlines, _git add_ displays the following:
```
$ git add README-crlf
warning: CRLF will be replaced by LF in airtime_mvc/public/images/doc/README-crlf.
The file will have its original line endings in your working directory.
```

This way if a user running Windows Git checks in a new or modified file with CR+LF newlines, Git will convert the newlines to LF upon checkin.

I also tried _git add_ on a new .png and no warning was given.

So, although it took a while, I think I found the simplest possible solution that works.

I'll submit a PR later on adding a .gitattributes file.

**Comment by hairmare** [comment]
@bburton Thanks for figuring this one out. Looking forward to your PR!

**Comment by Robbt** [comment]
For some reason after changing hard drives and moving my file system I ran into this after doing a pull.
A few js files most of the .MD files and all of the binary files were being marked as modified with the warning: CRLF will be replaced by LF in docs/manual/using-monit/index.md.

Was there a commit onto the docs from a windows box that did this or was it a file system issue on my own box. I haven't tested it fully on my other machine as I'm trying to consolidate my development on my laptop. Just thought I'd get some feedback. I currently have these files pending a commit as I was attempting to fix this and I want to get back to contributing now that my schedule is a little less cramped.

**Comment by hairmare** [comment]
The .md files have CRLF line terminators on my macOS box as well. 

It looks like this is also why the diff from #188 looked wrong. `docs/manual/upgrading/index.md` from there still seems to have CRLFs 😞. 

I'm not sure where all those CRLFs are coming from, it seems I added the .md ones during converting the legacy upstream html docs. I have no idea why some of the js files are wrong.

As far as I can tell the `.gitattributes` is somehow interfering with files that were committed with the wrong line terminators.

**Comment by bburton** [comment]
@Robbt @hairmare I was under the impression that Linux and OS-X git clients did not do text file end-of-line conversions as the Windows client does. However, upon re-reading the [.gitattributes manpage](https://git-scm.com/docs/gitattributes), it appears that clients on all platforms do text file end-of-line conversions.

If you look back at my long comment on April 6th, I tested adding a file with CR+LF line endings and the result is Git displayed a warning that the newlines were incompatible. However, I did not test a merge or what happens under Linux.

After the `.gitattributes` was added, it forced all files Git thinks are text files to retain the LF line endings when being checked out even when Windows Git is used as otherwise the text files would all be converted to have CR+LF's in the workspace. It may be that although it's forcing LF's on check-out, it might not convert CR+LF files upon commit or merge since the assumption is that all text files already have LF endings in the workspace. 

So the current `.gitattributes` file defaults to assuming all text files should not have their newlines converted even on Windows. In contrast, the following assumes all text files should have their newlines converted but then specifically overrides certain files and directories to retain LF endings so Vagrant provisioning can work even when the repository was cloned on Windows.

```
# Auto-detect text vs. binary files and ensure newlines are converted on 
# text files for checkouts, commits and merges.
* text=auto
# disable newline conversion on checkout with no conversion on commit
installer/* -text
dev_tools/* -text
# all filenames and extensions that must retain LF line endings
install text eol=lf
uninstall text eol=lf
*.sh text eol=lf
```

The above is the same as the originally proposed .gitattributes file above but some of the comments have been updated and binary files are not specified as the text=auto should detect them.

Once the `.gitattributes` file has been adjusted, any affected text files in the repository need to be adjusted. This page on [GitHub Help](https://help.github.com/articles/dealing-with-line-endings/#platform-linux) outlines how to do this.

**Comment by hairmare** [comment]
@bburton Not converting the installer for vagrant makes sense, I didn't consider it before.

Fixing .gitattributes seems to be the way forward. It makes more sense to have `vagrant up` work on windows than to guard against having newlines in files. 

It does look like .gittatributes isn't supported by all ui clients (https://superuser.com/a/882929). Mine does support it but it can behave funky when something unexpected is in the repo.

It looks like we'll have to take care when reviewing changes to non converted files made by windows users. The new gitattributes will still help with editing most parts of LibreTime. I tested the new gitattributes from above in a local clone, it seems to behave as intended. I don't have a windows box for testing vagrant with the new file though.

@Robbt I'm not quite sure where the markdown files got their CRLF line endings from. It looks like they come from my mac though. I've done some experimenting and am preparing a fix for the doc files. 

The affected js files all seem to be 3rd party libraries, maybe those had their CRLF endings from the get go. Given that the list of them is rather small I'm considering not touching them until we have to (ie. plupload should get bumped to a new version anyhow and we can fix it then). For the reference, the following js files are affected.

* airtime_mvc/public/js/contextmenu/jquery.contextMenu.js
* airtime_mvc/public/js/libs/jquery.stickyPanel.js
* airtime_mvc/public/js/plupload/i18n/hr_HR.js
* airtime_mvc/public/js/plupload/i18n/sr_RS.js
* airtime_mvc/public/js/plupload/i18n/sr_RS@latin.js
* airtime_mvc/public/js/sprintf/sprintf-0.7-beta1.js
* airtime_mvc/public/js/timepicker/jquery.ui.timepicker.js

I didn't find any binaries that exhibit the problem but I don't think I looked hard enough.

Anyhow, I'm solving the CRLF issue in `docs/` by rewriting them all with #202.

Maybe we should re-open this until `vagrant up` on Windows is fixed?
